sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -1056,23 +1054,43 @@ def deserialize_rows(self, read_rows_response, reader):
 
   def read(self, range_tracker):
     storage_client = bq_storage.BigQueryReadClient()
-    read_rows_iterator = iter(storage_client.read_rows(self.read_stream_name))
-    read_rows_response = next(read_rows_iterator)
-    avro_schema = avro.schema.Parse(read_rows_response.avro_schema.schema)
-    reader = avroio.DatumReader(avro_schema)
-    while True:
-      # self.deserialize_rows(read_rows_response, reader)
-      decoder = avroio.BinaryDecoder(
-          io.BytesIO(read_rows_response.avro_rows.serialized_binary_rows))
-      current_row = 0
-      while current_row < read_rows_response.row_count:
-        self.deserialized_rows.append(reader.read(decoder))
-        current_row += 1
-      read_rows_response = next(read_rows_iterator, None)
-      if read_rows_response is None:
-        break
-
-    return iter(self.deserialized_rows)
+    row_reader = _ReadRowsResponseReader(
+        storage_client.read_rows(self.read_stream_name))
+    return iter(row_reader)
+
+
+class _ReadRowsResponseReader():
+  def __init__(self, read_rows_responses):
+    self.read_rows_iterator = iter(read_rows_responses)
+    self.read_rows_response = next(self.read_rows_iterator)
+    self.avro_schema = avro.schema.Parse(
+        self.read_rows_response.avro_schema.schema)
+    self.reader = avroio.DatumReader(self.avro_schema)
+    self.decoder = avroio.BinaryDecoder(
+        io.BytesIO(self.read_rows_response.avro_rows.serialized_binary_rows))
+    self.current_row = 0
```

Comment:
Nit: this looks like the offset of the next row, in practice.

Reply:
Done